RHOAIENG-3459: Implement images cmd

devguyio opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
>[!WARNING]
> This PR is stacked on top of #234  by adding one extra commit. 
> Please review #234 first and focus the review here on the last commit only.

## Description
<!--- Describe your changes in detail -->
Add an images command to list & build model images.
    
This PR implement following odh subcommands:
    
- images: Lists model images from the model registry.
- images update: Updates the version image build parameters
- images build: Creates a new pipelinerun using the build params synced and stored in the model registry.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
```shell
cd cli
make cli-build
./odh images
```
![2024-04-22_15-05](https://github.com/opendatahub-io/ai-edge/assets/1930204/5e02cb06-d452-4524-bf12-86445657b9de)

```shell
./odh images update -i <model-id> -v <version-name>
```
![2024-04-29_20-31](https://github.com/opendatahub-io/ai-edge/assets/1930204/f7d0bb51-df37-47be-9c9d-f1affc6c0899)

```shell
./odh images build -i <model-id> -v <version-name>
```
![2024-04-29_20-40](https://github.com/opendatahub-io/ai-edge/assets/1930204/69da8226-4667-4526-acdb-0f05440ca205)

## Merge criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

devguyio commented:
@biswassri FYI

devguyio commented:
/cc @biswassri

grdryn commented:
/remove-approve

(new process)

Sara4994 commented on cli/pkg/commands/images/sync.go:
```diff
@@ -0,0 +1,47 @@
+	"github.com/spf13/cobra"
+)
+
+var syncCmd = common.NewCmd(
```
@devguyio As discussed this sometime back, In sync command can we include flags "imageId" and "version" then assign values to it. Directly passing values as arguments looks unclear on what exactly those values are.

devguyio replied:
Done

Sara4994 commented on cli/pkg/httptest/server.go:
```diff
@@ -108,3 +108,53 @@ func (m *MockServer) Close() {
 	return m.server.URL
 }
+
+// func (m *MockServer)
```
do we still need this commented out lines?

devguyio replied:
Removed

Sara4994 commented:
@devguyio also, can we include instructions about models sync and images list commands in docs [here](https://github.com/opendatahub-io/ai-edge/blob/main/docs/cli.md)

devguyio commented:
@Sara4994 @MarianMacik 

I've updated the documentation also the PR description with the new commands' syntax. I also updated the screenshot. 

- The images list doesn't have "sync" column anymore, since there's no need for that anymore.
- The sync subcommand is now called update.
- The sync and build subcommands use -i <model-id> -v <version-name> flags to identify the version image.

grdryn commented on cli/pkg/edgeclient/client.go:
```diff
@@ -96,13 +102,370 @@ func (c *Client) AddNewModelWithImage(
+	pipelineRun := &tektonv1.PipelineRun{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:    namespace,
+			GenerateName: "aiedge-e2e-",
```
It might make sense for this to include the model name also. We've started doing that with our example PipelineRuns, e.g.: https://github.com/opendatahub-io/ai-edge/blob/main/pipelines/tekton/aiedge-e2e/aiedge-e2e.tensorflow-housing.pipelinerun.yaml#L7

devguyio replied:
fixed

grdryn commented on cli/pkg/edgeclient/client.go:
```diff
@@ -96,13 +102,370 @@ func (c *Client) AddNewModelWithImage(
+							AccessModes: []corev1.PersistentVolumeAccessMode{corev1.ReadWriteOnce},
+							Resources: corev1.VolumeResourceRequirements{
+								Requests: corev1.ResourceList{
+									corev1.ResourceStorage: resource.MustParse("1Gi"),
```
Should that 1Gi be configurable somehow? Not necessarily in this PR, but in general I think it would be useful.

devguyio replied:
Probably we should do that in a separate task yes.

grdryn commented on cli/pkg/edgeclient/client.go:
```diff
@@ -96,13 +102,370 @@ func (c *Client) AddNewModelWithImage(
+	config, _ := clientcmd.BuildConfigFromFlags("", kubeConfig)
+	tektonClient, _ := tektonclientset.NewForConfig(config)
+
+	createdPipelineRun, err := tektonClient.TektonV1().PipelineRuns(namespace).Create(
```
Having the earlier "calculation" of constructing the PipelineRun object in a separate function than this "action" of sending it to the server might be a useful way to break it down a little? It should make testing of the different conditions in the construction of the object easier to test, without needing to worry about the server, for example.

devguyio replied:
Yeah makes sense, refactore.

grdryn commented on cli/pkg/modelregistry/client.go:
```diff
@@ -366,3 +566,9 @@ func encodeToBase64(v interface{}) (string, error) {
+func decodeFromBase64(v interface{}, enc string) error {
+	decoder := base64.NewDecoder(base64.StdEncoding, strings.NewReader(enc))
+	return json.NewDecoder(decoder).Decode(v)
+	// return json.NewDecoder(base64.NewDecoder(base64.StdEncoding, strings.NewReader(enc))).Decode(v)
```
```suggestion
```

devguyio replied:
fixed

grdryn reviewed (commented):
There are a few unresolved inline comments from @MarianMacik throughout here too.

Sara4994 commented:
Tested images and its subcommands locally along with Model-registry, they are working as expected. Will wait until the comments here are taken care of, before approving.

Sara4994 commented:
Tested images and its subcommands locally along with Model-registry, they are working as expected. Will wait until the comments here are taken care of, before approving.

devguyio commented on cli/pkg/edgeclient/client.go:
```diff
@@ -0,0 +1,537 @@
+					}
+					if a.ModelArtifact != nil && a.ModelArtifact.GetExternalID() != "" {
+						i.Id = a.ModelArtifact.GetExternalID()
+						i.NeedsSync = needsSync(m, v, a)
```
I removed the sync functionality as we agreed. Registered model & version are created with the `models add` command.

devguyio commented on cli/pkg/modelregistry/client.go:
```diff
@@ -286,6 +539,28 @@ func ToMetadataValueMap(
+			if err := decodeFromBase64(&sl, v.MetadataStructValue.GetStructValue()); err != nil {
+				return nil, fmt.Errorf("failed to decode metadata value for %s: %w", k, err)
+			}
+			props[k] = sl.Items
```
Correct and that's by design as the []interface{} is coming from the yaml decoding, which can't determine what's the type of a json array. And since we only support an [] of strings, so we only store `[] string` in MR, and return it as `[] string` for ease of later use.

LaVLaS commented on cli/pkg/commands/images/images.go:
```diff
@@ -0,0 +1,228 @@
+var Cmd = common.NewCmd(
+	"images",
+	"Manage model container images",
+	`Manage Open Data Hub model container images from the command line.
+
+This command allows you to list and build model container images suitable for deployment in edge environments.`,
```
Did we move away from Inference Container image?  With the ongoing discussion about using OCI containers images for model storage, "model container images" could be confused with that initiative when we are focused on providing a container that provides an inferencing service (model + serving runtime)

devguyio replied:
Changed to `inference container image`

Sara4994 commented:
@devguyio any updates on this PR?

devguyio commented on cli/pkg/edgeclient/client.go:
```diff
@@ -96,13 +102,370 @@ func (c *Client) AddNewModelWithImage(
+		testDataConfigMapName = tdc
+	}
+
+	if adw, ok := parameters["addDirWorkspace"]; !ok {
+		return nil, fmt.Errorf("addDirWorkspace pipeline parameter is required")
+	} else if adw, ok := adw.(string); !ok {
+		return nil, fmt.Errorf("addDirWorkspace pipeline parameter must be a boolean")
+	} else {
+		addDirWorkspace = adw == "true"
+	}
```
This is actually still available in [the bike-rental app `PipelineRun`](https://github.com/opendatahub-io/ai-edge/blob/main/pipelines/tekton/aiedge-e2e/aiedge-e2e.bike-rentals.pipelinerun.yaml#L59-L60), it's why it's conditional

devguyio commented on cli/pkg/edgeclient/client.go:
```diff
@@ -74,20 +82,18 @@ func (c *Client) AddNewModelWithImage(
 	m, v, a, err := c.modelRegistryClient.AutoRegisterModelVersionArtifact(
-		modelName, modelDescription, modelVersion, modelName, externalID, uri, modelFormatName, "", md,
+		modelName, modelDescription, modelVersion, modelName, uri, modelFormatName, "", md,
 	)
 	if err != nil {
 		return nil, fmt.Errorf("failed to add model image: %w", err)
 	}
 
 	return &ModelImage{
-		ID:          a.ModelArtifact.GetExternalID(),
 		ModelID:     m.GetId(),
```
The reason I used this initially was to be able to look up an artifact by a globally unique ID as that doesn't exist  in MR. But since I've moved the UX to just identify an artifact by the `--model-id <id> --version <version>`  , then that globally unique ID is no longer needed.

devguyio commented on cli/pkg/modelregistry/client_test.go:
```diff
@@ -128,6 +128,7 @@ func compareRegisteredModels(t *testing.T, model openapi.RegisteredModel, expect
 	if model.GetDescription() != expected.GetDescription() {
 		t.Fatalf("\t%s:\t%s\tShould receive the expected Description, got %s", t.Name(), failed, model.GetDescription())
 	}
+
```
Done

devguyio commented on cli/pkg/edgeclient/client.go:
```diff
@@ -96,13 +102,370 @@ func (c *Client) AddNewModelWithImage(
+		s3SecretName = s3s
+	}
+
+	if tdc, ok := parameters["testDataConfigMapName"]; !ok {
+		return nil, fmt.Errorf("testDataConfigMapName pipeline parameter is required")
+	} else if tdc, ok := tdc.(string); !ok {
+		return nil, fmt.Errorf("testDataConfigMapName pipeline parameter must be a string")
+	} else {
+		testDataConfigMapName = tdc
+	}
```
Done

devguyio commented on cli/pkg/commands/models/models.go:
```diff
@@ -114,8 +120,7 @@ func (m teaModel) View() string {
 	case common.SubCommandAdd:
-		return common.MessageStyle.Render("\nModel added successfully\n\n")
-
+		return common.MessageStyle.Render("\nModel information added successfully\n\n")
```
Done

devguyio commented:
> Just repeating this to not get it lost.
> This shouldn't be needed I think. If there are no versions, the next range will be noop, right?

@MarianMacik yes, nice catch! Fixed. I also addressed all your other comments.

devguyio commented on cli/pkg/commands/common/styles.go:
```diff
@@ -18,6 +18,13 @@ package common
+	// CheckMark is the checkmark symbol
+	CheckMark = "✓"
+	// WaringSymbol is the warning symbol
+	WaringSymbol = "⚠️"
```
Fixed, the symbols were removed an they're no longer used.

devguyio commented on cli/pkg/commands/images/update.go:
```diff
@@ -0,0 +1,49 @@
+		flags.FlagParams.Shorthand(),
+	),
+	"Update an edge model image parameters in the model registry.",
+	`Update an edge model image parameters in the model registry.
```
fixed

devguyio commented on cli/pkg/commands/models/add.go:
```diff
@@ -17,20 +17,45 @@ limitations under the License.
+	),
+	"Add a model information to the model registry including the model name, model description, model version "+
+		"and build parameters.",
+	`Add a model image information to the model registry including the model name, model description, model version and build parameters.
```
Fixed

devguyio commented on cli/pkg/edgeclient/client.go:
```diff
@@ -96,13 +102,370 @@ func (c *Client) AddNewModelWithImage(
+	artifact, err := c.modelRegistryClient.FindModelVersionArtifact(modelVersionID, artifactName)
+	if err != nil {
+		if errors.Is(err, modelregistry.ErrFindArtifact) {
+			// TODO: Figure out what to do with the URI
```
This is something that we probably need to work on in the future. Maybe add some logic in the pipeline to update that URL with the container image URL.

devguyio commented on cli/pkg/edgeclient/client.go:
```diff
@@ -96,13 +102,370 @@ func (c *Client) AddNewModelWithImage(
+		},
+	}
+
+	if addDirWorkspace {
+		pipelineRun.Spec.Workspaces = append(
+			pipelineRun.Spec.Workspaces,
+			tektonv1.WorkspaceBinding{
+				Name:     "workspace",
+				EmptyDir: &corev1.EmptyDirVolumeSource{},
+			},
+		)
+	}
```
This is actually still available in [the bike-rental app PipelineRun](https://github.com/opendatahub-io/ai-edge/blob/main/pipelines/tekton/aiedge-e2e/aiedge-e2e.bike-rentals.pipelinerun.yaml#L59-L60), it's why it's conditional

MarianMacik replied:
But this is just a PipelineRun. The Pipeline itself makes no use of that workspace:
https://github.com/opendatahub-io/ai-edge/blob/3d655f307b4a4648db7b7df56be74a43220d4fdb/pipelines/tekton/aiedge-e2e/aiedge-e2e.pipeline.yaml#L375-L384
I have opened a PR for it in #258.
You can remove this here.

devguyio replied:
@MarianMacik removed

devguyio commented:
@Sara4994 @MarianMacik @grdryn I've addressed all the comments.

devguyio commented:
> @Sara4994 @MarianMacik @grdryn I've addressed all the comments.

I've also updated the screenshots in the PR description.

MarianMacik reviewed (commented):
Looks good, I will put lgtm once the `addDirWorkspace` is removed as it is not needed.

devguyio commented:
> Looks good, I will put lgtm once the `addDirWorkspace` is removed as it is not needed.

Removed

MarianMacik reviewed (approved):
/lgtm

grdryn commented on cli/pkg/edgeclient/client.go:
```diff
@@ -96,13 +102,370 @@ func (c *Client) AddNewModelWithImage(
+	return images, nil
+}
+
+// UpdateModelImage synchronizes an edge model image information with the model registry by ensuring that the model
```
The bit that I think doesn't read great here is that it says 
> [...] synchronizes an _____ ______ information with the model registry [...]
where I think the following is better:
> [...] synchronizes ______ information with the model registry [...]

"an information" doesn't sound right.

devguyio replied:
Done

grdryn reviewed (approved):
/lgtm with some questions/comments inline

devguyio commented on cli/pkg/modelregistry/client.go:
```diff
@@ -284,33 +408,87 @@ func (c *Client) FindRegisteredModel(name string) (*openapi.RegisteredModel, err
+	}
+	if newMetaData == nil {
+		return v, nil
+	}
```
No, that was a left over from a former version of that method. Removed.

devguyio commented on cli/pkg/edgeclient/client.go:
```diff
@@ -96,13 +102,370 @@ func (c *Client) AddNewModelWithImage(
+			}
+			if len(artifacts) > 0 {
+
+				// TODO: Set the status based on whether the image is built or not
+				// TODO: Figure out where to show the image SHA
+				// TODO: Figure out the URI
```
It's not expected as part of this PR, JIRA created and mentioned.

devguyio commented on cli/pkg/modelregistry/client.go:
```diff
@@ -255,6 +252,133 @@ func (c *Client) GetRegisteredModels() ([]openapi.RegisteredModel, error) {
+		// ErrModelNotFound if the error message contains the expected error message.
+		// TODO: Remove this workaround when model registry returns 404 when the model is not found
+		if resp.StatusCode != 200 && isOpenAPIErrorOfKind(err, ErrModelNotFound) {
+			return nil, fmt.Errorf("%w1d: %s", ErrModelNotFound, id)
```
That was a typo, that should've been `%w. model id: %s` . Fixed.

devguyio commented:
> /lgtm with some questions/comments inline

@grdryn comments addressed. You can view the changes [here](https://github.com/opendatahub-io/ai-edge/compare/a360103411e0b8eea7ad76cd33e3826b1177cf40..3eb130b66bcc5af2d433a31f4c05c24c8d4f873a)

grdryn reviewed (approved):
/approve

MarianMacik reviewed (approved):
/lgtm
